plugins/builds/steps/update.js

```diff
@@ -0,0 +1,62 @@
+'use strict';
+const boom = require('boom');
+const schema = require('screwdriver-data-schema');
+
+module.exports = () => ({
+    method: 'PUT',
+    path: '/builds/{id}/steps/{name}',
+    config: {
+        description: 'Update a step for a build',
+        notes: 'Writes a step record',
+        tags: ['api', 'builds', 'steps'],
+        auth: {
+            strategies: ['token'],
+            scope: ['build']
+        },
+        handler: (request, reply) => {
+            const factory = request.server.app.buildFactory;
+            const buildId = request.params.id;
+            const buildIdCred = request.auth.credentials.username;
+            let stepIndex = -1;
+
+            if (buildId !== buildIdCred) {
+                throw boom.forbidden(`Credential only valid for ${buildIdCred}`);
+            }
+
+            return factory.get(buildId)
+                .then(build => {
+                    if (!build) {
+                        throw boom.notFound('Build does not exist');
+                    }
+                    const steps = build.steps;
+                    const now = (new Date()).toISOString();
+
+                    stepIndex = steps.findIndex((step) => step.name === request.params.name);
+
+                    if (stepIndex === -1) {
+                        throw boom.notFound('Step does not exist');
+                    }
+
+                    if (request.payload.code) {
+                        steps[stepIndex].code = request.payload.code;
+                        steps[stepIndex].endTime = request.payload.endTime || now;
+                    } else {
+                        steps[stepIndex].startTime = request.payload.startTime || now;
```

Comment:
It still kinda concerns me that our validation is not very strict. Like they can put endTime only in the payload. And then this would update the startTime instead.